src/systemcmds/topic_listener/listener_main.cpp

```diff
@@ -100,62 +103,54 @@ void listener(listener_print_topic_cb cb, const orb_id_t &id, unsigned num_msgs,
 		int sub = orb_subscribe_multi(id, topic_instance);
 		orb_set_interval(sub, topic_interval);
 
-		bool updated = false;
 		unsigned i = 0;
-		hrt_abstime start_time = hrt_absolute_time();
 
 		while (i < num_msgs) {
 
-			// check for user input to quit
-			int user_input_timeout = 1;
-
-			orb_check(sub, &updated);
-
-			if (i == 0) {
-				updated = true;
-				user_input_timeout = 0;	// don't wait
-			}
-
-			// check for user input
-			struct pollfd fds {};
-			fds.fd = 0; /* stdin */
-			fds.events = POLLIN;
-
-			if (poll(&fds, 1, 0) > 0) {
-
-				char c = 0;
-				int ret = read(0, &c, user_input_timeout);
-
-				if (ret) {
-					return;
-				}
-
-				switch (c) {
-				case 0x03: // ctrl-c
-				case 0x1b: // esc
-				case 'q':
-					return;
-					/* not reached */
+			struct pollfd fds[2] {};
+			// Poll for user input (for q or escape)
+			fds[0].fd = 0; /* stdin */
+			fds[0].events = POLLIN;
+			// Poll the UOrb subscription
+			fds[1].fd = sub;
+			fds[1].events = POLLIN;
+
+			if (poll(&fds[0], 2, int(MESSAGE_TIMEOUT_S * 1000)) > 0) {
+
+				// Received character from stdin
+				if (fds[0].revents & POLLIN) {
+					char c = 0;
+					int ret = read(0, &c, 1);
+
+					if (ret) {
+						return;
+					}
+
+					switch (c) {
+					case 0x03: // ctrl-c
+					case 0x1b: // esc
+					case 'q':
+						return;
+						/* not reached */
+					}
 				}
-			}
 
-			if (updated) {
-				start_time = hrt_absolute_time();
-				i++;
+				// Received message from subscription
+				if (fds[1].revents & POLLIN) {
+					i++;
```

Comment:
While you're at it, can you give this `i` a better name? E.g. `num_messages_printed`?